### PR TITLE
fix cross-tenant subscription related issues

### PR DESCRIPTION
**Pull request by tharindu1st**
resolves #https://github.com/wso2/product-apim/issues/11640

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java [hunk 1]
```diff
@@ -989,4 +923,11 @@ String getAsyncAPIDefinitionForLabel(Identifier apiId, String labelName)
      * @throws APIManagementException if failed to get the related API revision Deployment Mapping details
      */
     List<APIRevisionDeployment> getAPIRevisionDeploymentListOfAPI(String apiUUID) throws APIManagementException;
+
+    Set<SubscribedAPI> getPaginatedSubscribedAPIsByApplication(Application application, Integer offset, Integer limit
+            , String organization) throws APIManagementException;
+
+    List<Tier> getThrottlePolicies(int tierApiType, String organization) throws APIManagementException;
+
+    Tier getThrottlePolicyByName(String policyId, int policyType, String organization) throws APIManagementException;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java:932]: let's add java docs.

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/SubscribedAPI.java [hunk 2]
```diff
@@ -59,6 +59,17 @@ public SubscribedAPI(String uuid) {
         this.uuid = uuid;
     }
 
+
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/SubscribedAPI.java:62]: extra new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 3]
```diff
@@ -5989,6 +5899,79 @@ public String getAsyncAPIDefinitionForLabel(Identifier apiId, String labelName)
...
+                subscribedAPIs = apiMgtDAO.getPaginatedSubscribedAPIsByApplication(application, offset,
+                    limit, organization);
+            if (subscribedAPIs != null && !subscribedAPIs.isEmpty()) {
+                Map<String, Tier> tiers = APIUtil.getTiers(tenantId);
+                for (SubscribedAPI subscribedApi : subscribedAPIs) {
+                    Tier tier = tiers.get(subscribedApi.getTier().getName());
+                    subscribedApi.getTier().setDisplayName(tier != null ? tier.getDisplayName() : subscribedApi
+                            .getTier().getName());
+                    // We do not need to add the modified object again.
+                    // subscribedAPIs.add(subscribedApi);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:5918]: let's remove these.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 4]
```diff
@@ -5989,6 +5899,79 @@ public String getAsyncAPIDefinitionForLabel(Identifier apiId, String labelName)
...
+                    tiers.put(tierPermission.getTierName(), tier);
+                }
+            }
+
+            // Removing denied Tiers
+            Set<String> deniedTiers = getDeniedTiers();
+            for (String tierName : deniedTiers) {
+                tiers.remove(tierName);
+            }
+        }else if (tierLevel == APIConstants.TIER_APPLICATION_TYPE){
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:5947]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 5]
```diff
@@ -5989,6 +5899,79 @@ public String getAsyncAPIDefinitionForLabel(Identifier apiId, String labelName)
...
+            if (APIUtil.isOnPremResolver()){
+                organization = tenantDomain;
+            }
+            tiers = APIUtil.getTiers(APIConstants.TIER_APPLICATION_TYPE, organization);
+        }
+        if (tiers!= null){
+            tierList.addAll(tiers.values());
+        }
+        return tierList;
+    }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:5957]: Needs to be formatted

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 6]
```diff
@@ -7907,6 +7894,81 @@ public static String getDefaultThrottlingPolicy(String tenantDomain) {
...
+                    tier.setRateLimitCount(subscriptionPolicy.getRateLimitCount());
+                    tier.setRateLimitTimeUnit(subscriptionPolicy.getRateLimitTimeUnit());
+                    setBillingPlanAndCustomAttributesToTier(subscriptionPolicy, tier);
+                    if (StringUtils.equals(subscriptionPolicy.getBillingPlan(), APIConstants.COMMERCIAL_TIER_PLAN)) {
+                        tier.setMonetizationAttributes(subscriptionPolicy.getMonetizationPlanProperties());
+                    }
+                }
+
+                if (limit instanceof RequestCountLimit) {
+
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:7934]: remove extra new line

**Review by bhathiya — COMMENTED** [review]
Added a few comments

**Review by bhathiya — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 7]
```diff
@@ -484,7 +484,9 @@
     SERVICE_VERSION_NOT_FOUND(901900, "Cannot find the service version", 404, "Cannot find a service that matches the given version"),
     INVALID_ENDPOINT_CREDENTIALS(902000, "Invalid Endpoint Security credentials", 400,
             "Invalid Endpoint Security credentials. %s", false),
-    INVALID_TENANT_CONFIG(9020001, "Invalid tenant-config found", 400, "Invalid tenant-config found with error %s", false);
+    INVALID_TENANT_CONFIG(9020001, "Invalid tenant-config found", 400, "Invalid tenant-config found with error %s", false),
+    SUBSCRIPTION_TIER_NOT_ALLOWED(9020002, "Subscription Tier is not allowed for user", 403, "Subscription Tier %s is" +
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:488]:
```suggestion
    INVALID_TENANT_CONFIG(902001, "Invalid tenant-config found", 400, "Invalid tenant-config found with error %s", false),
    SUBSCRIPTION_TIER_NOT_ALLOWED(902002, "Subscription Tier is not allowed for user", 403, "Subscription Tier %s is" +
```

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**